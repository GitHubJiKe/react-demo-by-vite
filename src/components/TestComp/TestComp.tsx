import React, { useState } from "react";

interface TestCompProps {
  data: number[];
  onClick(count: number): void;
}

function TestComp({ data, onClick }: TestCompProps) {
  const [count, setCount] = useState(0);
  return (
    <div
      onClick={() => {
        setCount((c) => c + 1);
        onClick && onClick(count);
      }}
    >
      <div className="count">{count}</div>
      {data.map((v) => (
        <Foo val={v} key={v} />
      ))}
    </div>
  );
}

export function Foo({ val }: { val: number }) {
  return <div>Foo {val}</div>;
}

export default TestComp;
