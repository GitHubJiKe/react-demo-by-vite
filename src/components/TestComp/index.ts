export { default } from './TestComp';
export * from './TestComp';
