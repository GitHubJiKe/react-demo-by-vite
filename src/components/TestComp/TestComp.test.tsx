import React from "react";
import TestComp, { Foo } from "./TestComp";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import sinon from "sinon";

const data = [1, 2, 3, 4];
const onClick = sinon.spy();

describe("render a TestComp and click to change count", () => {
  const comp = shallow(<TestComp data={data} onClick={onClick} />);
  comp.simulate("click"); // 模拟点击事件
  comp.simulate("click");
  comp.simulate("click");

  it("renders an `.count`", () => {
    //   查找组件内的某个子节点存在与否
    expect(comp.find(".count")).toHaveLength(1);
  });

  it("count value equals 3", () => {
    //   以及判断子节点的内容是什么
    expect(comp.find(".count").text()).toEqual("3");
  });

  it(`TestComp should has ${data.length} Foo Child`, () => {
    // 组件应该有多少个某个子组件
    expect(comp.find(Foo)).toHaveLength(data.length);
  });

  it("onClick func has been called 3 times", () => {
    //   某事件被调用了多少次
    expect(onClick).toHaveProperty("callCount", 3);
  });
});

test("TestComp should be these:", () => {
  const renderComp = renderer.create(
    <TestComp data={data} onClick={onClick} />
  );

  const json = renderComp.toJSON();

  expect(json).toMatchSnapshot();
});
