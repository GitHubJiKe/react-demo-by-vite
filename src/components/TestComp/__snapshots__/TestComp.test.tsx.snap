// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestComp should be these: 1`] = `
<div
  onClick={[Function]}
>
  <div
    className="count"
  >
    0
  </div>
  <div>
    Foo 
    1
  </div>
  <div>
    Foo 
    2
  </div>
  <div>
    Foo 
    3
  </div>
  <div>
    Foo 
    4
  </div>
</div>
`;
